### ISSUE TITLE
fit_generator: loss and val_loss calculated differently, if same data is used for training and validation, different values?

### ISSUE THREAD

**Issue by ai-padawan** [keras-team/keras#12407]
There's something odd as to how val_loss is being computed.

Here's a sample MNIST code. Here, I am using the same generator, using the same dataset, and my loss and val_loss are not identical. This means I am using the same data to train and validate, and I should expect identical loss values, but it's not. Is there a bug in the code that calculates loss and val_loss differently?

> from keras.utils import np_utils
> from keras.models import Sequential
> from keras.layers import Dense, Dropout, Flatten, Activation, Conv2D, MaxPooling2D
> from keras.optimizers import Adam
> from keras.callbacks import EarlyStopping, ReduceLROnPlateau, ModelCheckpoint
> from keras.preprocessing.image import ImageDataGenerator
> import numpy as np # linear algebra
> import os
> from keras.datasets import mnist
> 
> (x_train, y_train), (x_val, y_val) = mnist.load_data()
> 
> x_train = x_train.astype("float32")/255.0
> x_val = x_val.astype("float32")/255.0
> print(y_train.shape)
> y_train = np_utils.to_categorical(y_train)
> y_val = np_utils.to_categorical(y_val)
> print(y_train.shape)
> 
> n_classes = y_train.shape[1]
> x_train = x_train.reshape(x_train.shape[0], 28, 28, 1)
> x_val = x_val.reshape(x_val.shape[0], 28, 28, 1)
> 
> print(x_train.shape, x_val.shape, y_train.shape, y_val.shape)
> 
> model = Sequential()
> 
> model.add(Conv2D(64,(3,3), input_shape=(28, 28, 1)))
> model.add(Activation('relu'))
> model.add(Conv2D(64,(3,3)))
> model.add(Activation('relu'))
> model.add(MaxPooling2D(pool_size=(2,2)))
> 
> model.add(Conv2D(128,(3,3)))
> model.add(Activation('relu'))
> model.add(Conv2D(128,(3,3)))
> model.add(Activation('relu'))
> 
> model.add(Flatten())
> model.add(Dense(1024))
> model.add(Activation('relu'))
> model.add(Dropout(0.5))
> model.add(Dense(10))
> model.add(Activation('softmax'))
> 
> datagen = ImageDataGenerator(zoom_range = 0.2,
>                             height_shift_range = 0.25,
>                             width_shift_range = 0.25,
>                             rotation_range = 25)
> model.compile(loss='categorical_crossentropy', optimizer = 'adam', metrics = ["accuracy"])
> 
> callbacks = [EarlyStopping(monitor='val_loss', min_delta=0, patience=10, verbose=2, mode='auto'),
>             ModelCheckpoint('mnist.h5', monitor='val_loss', save_best_only=True, verbose=0),
>             ReduceLROnPlateau(monitor='val_loss', factor=0.1, patience=3)]
>
> model.fit_generator(datagen.flow(x_train, y_train, batch_size = 512),
>                            steps_per_epoch = 10, 
>                            epochs = 10, 
>                            verbose = 1, 
>                            validation_data = datagen.flow(x_train, y_train, batch_size = 512),
>                            validation_steps = 10,
>                            callbacks = callbacks)

**Comment by iarganda** [comment]
I guess the augmented data generated on each call of datagen.flow are different, so that's why you get different values.
